druid-shell/src/platform/x11/application.rs

```diff
@@ -539,18 +574,15 @@ fn poll_with_timeout(conn: &Rc<XCBConnection>, idle: RawFd, timeout: Instant) ->
                     // There is an X11 event ready to be handled.
                     break;
                 }
+                let now = Instant::now();
+                if now >= idle_timeout.min(timer_timeout.unwrap_or(idle_timeout)) {
+                    break;
+                }
                 if poll_fds.len() == 1 || readable(poll_fds[1]) {
                     // Now that we got signalled, stop polling from the idle pipe and use a timeout
                     // instead.
                     poll_fds = &mut just_connection;
-
-                    let now = Instant::now();
-                    if now >= timeout {
-                        break;
-                    } else {
-                        poll_timeout = c_int::try_from(timeout.duration_since(now).as_millis())
-                            .unwrap_or(c_int::max_value())
-                    }
+                    poll_timeout = to_timeout(idle_timeout, now);
```

Comment:
... and then here you can keep
```suggestion
                    poll_timeout = to_timeout(idle_timeout, now);
                    if now >= timeout {
                        break;
                    }
```

Also, I think when resetting poll_timeout here you're forgetting the timer timeout. So maybe define `first_timeout` above the loop to be `idle_timeout.min(timer_timeout.unwrap_or(idle_timeout))`, and use `first_timeout` here.